### PR TITLE
Add documentation generator; add frost-p256 documentation using it

### LINKED ISSUE #95: Flesh out all rustdoc in frost-p256 public API
Half of the public items are well-doc'd, half just have placeholders to allow compilation.

### PR DESCRIPTION
**Pull request by conradoplg**
Since all ciphersuites documentation are basically the same, I added a Rust program to generate them from the canonical ristretto255 implementation. Using it, I added the P-256 docs.

Closes #95

### CODE REVIEW

**Comment by dconnolly** [comment]
A quick nit before diving in: any chance this could be written in Rust? 🙈

**Comment by conradoplg** [comment]
> A quick nit before diving in: any chance this could be written in Rust? see_no_evil

Why not? :rofl: 
I'll see if I can convert it without spending too much time on it

**Comment by conradoplg** [comment]
Done! :snake: :arrow_right:  :crab:

**Comment by dconnolly** [comment]
> Done! 🐍 ➡️ 🦀

Wow wow wow!

File: Cargo.toml [hunk 1]
```diff
@@ -6,3 +6,7 @@ members = [
     "frost-ristretto255",
     "frost-p256",
 ]
+
+# This is a development helper binary.
+# To run it, execute `cargo run --manifest-path gendoc/Cargo.toml`
+exclude = ["gendoc"]
```
[the remark this answers is not in the document]
Comment (dconnolly) [reply Cargo.toml:12]: We can make sure it doesn't get published while still keeping it maintained as part of the Rust workspace I think

Comment (dconnolly) [inline Cargo.toml:12]:
```suggestion
    "frost-p256",
    "gendoc"
]
```

File: gendoc/src/main.rs [hunk 2]
```diff
@@ -0,0 +1,123 @@
+//! Generate documentation for a ciphersuite based on another ciphersuite implementation.
+//!
+//! The documentation for each ciphersuite is very similar, with the only difference being
+//! the ciphersuite name.
+//!
+//! To make it easier to update all ciphersuites documentations when a change is needed,
+//! this binary allows updating all of them based on a single one. This binary
+//! uses frost-ristretto255 as the "canonical" one, so:
```
Comment (dconnolly) [inline gendoc/src/main.rs:8]:
```suggestion
//! To make it easier to update all ciphersuite documentation when a change is needed,
//! this binary allows updating all of them based on a single one. This binary
//! uses frost-ristretto255 as the "canonical" one, so:
```

File: gendoc/src/main.rs [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+//!
+//! - Change any documentation of a public function or struct in `frost-ristretto255/src/lib.rs`
+//! - Run `cargo run --manifest-path gendoc/Cargo.toml` to update the documentation
+//!   of the other ciphersuites.
+
+use std::{fs, iter::zip};
+
+use regex::Regex;
+
+/// Read the public documentation of public symbols in the given file.
```
Comment (dconnolly) [inline gendoc/src/main.rs:18]:
```suggestion
/// Read the public documentation of public types in the given file.
```

Comment (dconnolly) [reply gendoc/src/main.rs:18]: I think this is correct?

File: gendoc/src/main.rs [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+/// suite_names_code: strings that reference the specific suite in code
+///     inside `fn` and should be ignore when using for replacements.
+///
+/// # Returns
+///
+/// A list with data for each symbol, see above.
+fn read_docs(filename: &str, suite_names_code: &[&str]) -> Vec<(String, String, usize, usize)> {
+    let mut docs = Vec::new();
+    let code = fs::read_to_string(filename).unwrap();
+    let re = Regex::new(r"(?m)((^[ ]*///.*\n)+)\s*pub (.*)").unwrap();
```
Comment (dconnolly) [inline gendoc/src/main.rs:46]: I'm fine with regex'ing for now as this is an on-demand internal tool - if we are using it a lot, I wonder if there is some library API to hook into rustdoc processing itself, as the regex may be harder to maintain

File: Cargo.toml (hunk 1 above)
Comment (dconnolly) [reply Cargo.toml:12]: I won't just apply this as it makes `gendoc` the default executable when `cargo run` is done in the root of the workspace, which I'm not sure we want?

File: gendoc/src/main.rs (hunk 3 above)
Comment (conradoplg) [reply gendoc/src/main.rs:18]: I was looking for a word that covers types, functions, constants...

File: gendoc/src/main.rs (hunk 4 above)
Comment (conradoplg) [reply gendoc/src/main.rs:46]: Some time ago I tried this with `syn` but it's a bit annoying, it's not well documented, and I managed just to extract the docs, actually replacing them would be more difficult... so I went with the simplest thing that worked. But yeah, not a fan of regexes myself :sweat_smile:

File: Cargo.toml (hunk 1 above)
Comment (conradoplg) [reply Cargo.toml:12]: I'll try tinkering with the manifest to make it work

Comment (conradoplg) [reply Cargo.toml:12]: If we use `default-members` then `cargo run` won't run it, but you will need to use `cargo check --workspace` to check everything. What do you think?

File: gendoc/src/main.rs (hunk 3 above)
Comment (dconnolly) [reply gendoc/src/main.rs:18]: Perhaps 'items': https://doc.rust-lang.org/reference/items.html

Comment (conradoplg) [reply gendoc/src/main.rs:18]: Ah neat, thanks!

Comment (conradoplg) [reply gendoc/src/main.rs]: Replaced in aa95773

File: Cargo.toml (hunk 1 above)
[the remark this answers is not in the document]
Comment (dconnolly) [reply Cargo.toml]: Resolved for now with https://github.com/ZcashFoundation/frost/pull/130/commits/07ebaa3a1608a7acb96209f2ac80f06e264b6af1

Comment (dconnolly) [reply Cargo.toml]: Resolved for now with https://github.com/ZcashFoundation/frost/pull/130/commits/07ebaa3a1608a7acb96209f2ac80f06e264b6af1

**Review by dconnolly — APPROVED** [review]

**Comment by dconnolly** [comment]
@mergify rebase

**Review by dconnolly — DISMISSED** [review]

**Review by dconnolly — APPROVED** [review]

**Pull request merged**